### PR TITLE
Update bpds

### PR DESCRIPTION
**Pull request by counkomol**
This updates the patch sampler module (BufferedPatchDataset, BPDS), addressing the issues outlined in #130.

Some major changes include:

- Modification of the BPDS API to remove/rename misleading parameters.
- Update to the main training loop to pull batches of patch pairs from BPDS as opposed to using `torch`'s `DataLoader`. Reasoning:
  - BPDS is not a true dataset. Its index and length were meaningless.
  - BPDS has a global state that does not work with `DataLoader`'s multiprocessing.
- Fixes for some bugs/inefficiencies in BPDS.
- Reintroduction of a test to cover BPDS.

### CODE REVIEW

File: fnet/cli/train_model.py [hunk 1]
```diff
@@ -55,26 +54,17 @@ def init_cuda(gpu: int) -> None:
...
-    dataloader = torch.utils.data.DataLoader(bpds, batch_size=args.batch_size)
-    return dataloader
+    return fnet.data.BufferedPatchDataset(dataset=ds, **args.bpds_kwargs)
 
 
-def get_dataloader_val(
-        args: argparse.Namespace
-) -> Optional[torch.utils.data.DataLoader]:
-    """Creates DataLoader for validation."""
+def get_bpds_val(args: argparse.Namespace):
```
Comment (JacksonMaxfield) [inline fnet/cli/train_model.py]: Nit: no return type annotation on this function

Also without knowing too much as to what is happening, is it really best here to return optionally None? Why not throw an error? If the dataset val is None then I feel like a `NoDatasetError` or something should be thrown?

File: fnet/data/bufferedpatchdataset.py [hunk 2]
```diff
@@ -1,110 +1,182 @@
...
 import torch
 
-from fnet.data.fnetdataset import FnetDataset
-
 
 logger = logging.getLogger(__name__)
+ArrayLike = Union[np.ndarray, torch.Tensor]
+
 
+class BufferedPatchDataset:
```
Comment (JacksonMaxfield) [inline fnet/data/bufferedpatchdataset.py]: Swell commenting my dude.

File: fnet/data/bufferedpatchdataset.py [hunk 3]
```diff
@@ -1,110 +1,182 @@
...
-        patch_size=(32, 64, 64),
-        buffer_size=1,
-        buffer_switch_frequency=-1,
-        npatches=100000,
-        verbose=False,
-        shuffle_images=True,
-        dim_squeeze=None,
+            self,
+            dataset: collections.abc.Sequence,
+            patch_shape: Sequence[int] = (32, 64, 64),
```
Comment (JacksonMaxfield) [inline fnet/data/bufferedpatchdataset.py]: Huh, sequence type. Neat.

File: fnet/data/bufferedpatchdataset.py [hunk 4]
```diff
@@ -1,110 +1,182 @@
...
-
-    def __getitem__(self, index):
+
+        self.counter = 0
+        self.epochs = -1  # incremented to 0 when buffer initially filled
+        self.buffer = deque()
+        self.remaining_to_be_in_buffer = deque()
+        self.buffer_history = []
+        pbar = tqdm(range(self.buffer_size))
+        pbar.set_description('Buffering images')
```
Comment (JacksonMaxfield) [inline fnet/data/bufferedpatchdataset.py]: Could also have put this description in the tqdm init:

`pbar = tqdm(range(self.buffer_size), desc='Buffering images')`

File: fnet/data/bufferedpatchdataset.py [hunk 5]
```diff
@@ -1,110 +1,182 @@
...
+        for _ in pbar:
+            self.insert_new_element_into_buffer()
+
+    def __iter__(self):
```
Comment (JacksonMaxfield) [inline fnet/data/bufferedpatchdataset.py]: Does this just return the same object on iter? Shouldn't this be like `super().__iter__()` or something?

File: fnet/data/bufferedpatchdataset.py [hunk 6]
```diff
@@ -1,110 +1,182 @@
...
-    def insert_new_element_into_buffer(self):
-        # sample with replacement
-        self.buffer.pop(0)
-        if self.shuffle_images:
+        return patch
 
-            if len(self.remaining_to_be_in_buffer) == 0:
-                self.remaining_to_be_in_buffer = np.arange(
-                    0, len(self.dataset)
+    def _check_last_datum(self) -> None:
```
Comment (JacksonMaxfield) [inline fnet/data/bufferedpatchdataset.py]: @gregjohnso the singular of data is datum 🙃

File: fnet/cli/train_model.py (hunk 1 above)
Comment (counkomol) [reply fnet/cli/train_model.py]: I pushed an update to add some return type annotations. For training models, providing a validation dataset is optional, so this function returns either the validation dataset or None. The script checks the type later downstream.

File: fnet/data/bufferedpatchdataset.py (hunk 4 above)
Comment (counkomol) [reply fnet/data/bufferedpatchdataset.py]: Thanks for this suggestion. I'm not too familiar with `tqdm` or its API. I updated the code with your suggestion.

File: fnet/data/bufferedpatchdataset.py (hunk 5 above)
Comment (counkomol) [reply fnet/data/bufferedpatchdataset.py]: `BufferedPatchDataset` does not have a parent class so returning `super()`-something is not really relevant. The point of implementing `__iter__` is so users can do something like:

```python
bpds = BufferedPatchDataset(**some_kwargs)
for patch in bpds:
    process(patch)
```
Right now this is an infinite loop but might change in the future. e.g., stopping iteration after a specific number of patches or after a specific number of epochs through the input dataset.

Comment (JacksonMaxfield) [reply fnet/data/bufferedpatchdataset.py]: Ahhhh got it. Why not just make this your favourite thing then? `NotImplementedError` 🙃

Comment (counkomol) [reply fnet/data/bufferedpatchdataset.py]: Well, because right now it is implemented. It's just an infinite loop, which might be just fine for some people if their use their own stopping criteria. :P

File: fnet/data/bufferedpatchdataset.py [hunk 7]
```diff
@@ -1,110 +1,180 @@
...
+        """Checks last dataset item added to buffer."""
+        nd = len(self.patch_shape)
+        idx_buf = self.buffer_history[-1]
+        shape_spatial = None
+        for idx_c, component in enumerate(self.buffer[-1]):
+            if shape_spatial is None:
+                shape_spatial = component.shape[-nd:]
+            elif component.shape[-nd:] != shape_spatial:
```
Comment (heeler) [inline fnet/data/bufferedpatchdataset.py]: This block of code is a little hard to follow. Particularly the negative indexes in the slice notation. A comment or two would make this a lot more transparent.

File: tests/test_bufferedpatchdataset.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
+from collections import Counter
+from typing import Tuple
+
+import numpy as np
+import numpy.testing as npt
+import pytest
+
+from fnet.data import BufferedPatchDataset
+
+
```
Comment (heeler) [inline tests/test_bufferedpatchdataset.py]: This is VERY clean code to read! I really like it!

**Comment by heeler** [comment]
It's nicely written there are a few spots that could use more comments especially when you're jumping through loops with negative indexs etc but all in all great work.
